locale.c: Save output of emulate_setlocale in buffer

khwilliamson opened the pull request:
Depending on Configuration and platform and details of the current request, the value returned could be pointing to a system static buffer, or be a temporary freeable upon LEAVE.  This commit standardizes it to a known per-interpreter buffer that can be properly freed at termination.

khwilliamson committed:
locale.c: Save output of emulate_setlocale in buffer

Depending on Configuration and platform and details of the current
request, the value returned could be pointing to a system static buffer,
or be a temporary freeable upon LEAVE.  This commit standardizes it to a
known per-interpreter buffer that can be properly freed at termination.

tonycoz reviewed (commented):
I'm lost in a maze of macros, my main concerns are:
- whether it's possible for `save_to_buffer()` to be called with `buf == PL_stdize_locale_buf` (causing undefined behaviour in `Copy()`/`memcpy()`)
- whether the locale parameter can ever be `PL_stdize_locale_buf`, but then have the allocation `realloc()`ed out from under it.

khwilliamson committed:
locale.c: Teach save_to_buffer to handle self param

This function is called to save a string to a buffer.  Teach it to treat
as a no-op the string passed being the buffer itself.  This generalizes
it to make it work properly under more circumstances; the commit also
removes the current case where the function call was explicitly avoided
under this circumstance.

tonycoz commented:
I think the assert() is a good idea.

khwilliamson commented:
assert added
